cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/ComputeEngineCredentials.java

```diff
@@ -0,0 +1,124 @@
+  private static final Logger LOG = LoggerFactory.getLogger(ComputeEngineCredentials.class);
+  private static final Gson GSON = new Gson();
+  /**
+   * Time (in millisecond) to refresh the credentials before it expires.
+   */
+  private static final int NUMBER_OF_RETRIES = 10;
+  private static final int MIN_WAIT_TIME_MILLISECOND = 500;
+  private static final int MAX_WAIT_TIME_MILLISECOND = 10000;
+  private final String endPoint;
+
+  public ComputeEngineCredentials(@Nullable String endPoint) {
+    this.endPoint = endPoint;
+  }
+
+  private AccessToken refreshAccessTokenLocally() throws IOException {
+    try {
+      GoogleCredentials googleCredentials = com.google.auth.oauth2.ComputeEngineCredentials.create();
+      return googleCredentials.refreshAccessToken();
+    } catch (IOException e) {
+      throw new IOException("Unable to get credentials from the environment. "
+                              + "Please explicitly set the account key.", e);
+    }
+  }
+
+  private AccessToken refreshAccessTokenRemotely(String endPoint) throws IOException {
+    ExponentialBackOff backOff = new ExponentialBackOff.Builder()
+      .setInitialIntervalMillis(MIN_WAIT_TIME_MILLISECOND)
+      .setMaxIntervalMillis(MAX_WAIT_TIME_MILLISECOND).build();
+
+    Exception exception = null;
+    int counter = 1;
+    while (counter <= NUMBER_OF_RETRIES) {
+      try {
+        return fetchAccessToken(endPoint);
+      } catch (Exception ex) {
+        // exception does not get logged since it might get too spammy.
+        exception = ex;
+      }
+
+      try {
+        Thread.sleep(backOff.nextBackOffMillis());
+      } catch (InterruptedException ex) {
+        exception = ex;
+        break;
+      }
+      counter++;
+    }
+
+    LOG.error("Failed to fetch GoogleCredentials after {} retries", counter, exception);
+    return null;
+  }
+
+  private AccessToken fetchAccessToken(String endPoint) throws IOException {
+    URL url = new URL(endPoint);
+    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+    connection.connect();
+    try (Reader reader = new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8)) {
+      if (connection.getResponseCode() != HttpResponseStatus.OK.code()) {
+        throw new IOException(CharStreams.toString(reader));
+      }
+      GenericData token = GSON.fromJson(reader, GenericData.class);
+
+      String key = token.get("access_token").toString();
+      Double expiration = Double.parseDouble(token.get("expires_in").toString());
+      long expiresAtMilliseconds = System.currentTimeMillis() +
+        expiration.longValue() * 1000;
+
+      return new AccessToken(key, new Date(expiresAtMilliseconds));
+    } finally {
+      connection.disconnect();
+    }
+  }
+
+  @Override
+  public AccessToken refreshAccessToken() throws IOException {
+    //todo remove this after tests.
+    LOG.error(">>>>>>> calling refreshAccessToken");
```

Comment:
Remove?

Reply:
Yes, this will be removed once I'm done with my tests.